stan/math/rev/mat/fun/adj_jac_apply.hpp

```diff
@@ -18,34 +35,224 @@ namespace math {
+    offsets_[t] = count;
+    count += 1;
+    return count_memory(count, args...);
+  }
+
+  template <typename... Pargs>
+  size_t count_memory(size_t count, const double& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    offsets_[t] = count;
+    return count_memory(count, args...);
+  }
+
+  template <typename... Pargs>
+  size_t count_memory(size_t count, const int& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    offsets_[t] = count;
+    return count_memory(count, args...);
+  }
+
+  size_t count_memory(size_t count) { return count; }
+
+  /*
+   * prepare_x_vis recursively populates x_vis_ with the varis from each of its
+   * input arguments. The vari pointers for argument n are copied into x_vis_ at
+   * the index starting at offsets_[n].
+   *
+   * Each of the arguments can be an Eigen::Matrix with var or double scalar
+   * types, a std::vector with var, double, or int scalar types, or a var, a
+   * double, or an int.
+   *
+   * As an example, take the call:
+   *    prepare_x_vis(a1, a2, a3, a4, a5)
+   *
+   * If a1 is an Eigen::Matrix<var, 5, 5>, a2 is a double, a3 is a var, a4 is
+   * a length 5 vector of vars, and a5 is a length 3 vector of ints,
+   * x_vis_ to x_vis_ + 24 is filled with the vari pointers of a1,
+   * x_vis_[25] is set equal to the vari pointer of a3, x_vis_ + 26 to x_vis_ +
+   * 30 is filled with the vari pointers of a4. a2 and a5 have no varis and so
+   * are skipped.
+   *
+   * @param x next argument to have its vari pointers copied if necessary
+   * @param args the rest of the arguments (that will be iterated through
+   * recursively)
+   */
+  template <int R, int C, typename... Pargs>
+  void prepare_x_vis(const Eigen::Matrix<var, R, C>& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    for (int i = 0; i < x.size(); ++i) {
+      x_vis_[offsets_[t] + i] = x(i).vi_;
+    }
+    prepare_x_vis(args...);
+  }
+
+  template <int R, int C, typename... Pargs>
+  void prepare_x_vis(const Eigen::Matrix<double, R, C>& x,
+                     const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<var>& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    for (size_t i = 0; i < x.size(); ++i) {
+      x_vis_[offsets_[t] + i] = x[i].vi_;
+    }
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<double>& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<int>& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const var& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
```

Comment:
This is fiendishly clever, by the way (computing sizes this way).  I do worry a bit about compile times.